streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -45,59 +63,76 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
+                                 false,
+                                 streamPatternSourceNode);
     }
 
     @SuppressWarnings("unchecked")
     public <K, V> KTable<K, V> table(final String topic,
                                      final ConsumedInternal<K, V> consumed,
                                      final MaterializedInternal<K, V, KeyValueStore<Bytes, byte[]>> materialized) {
         final StoreBuilder<KeyValueStore<K, V>> storeBuilder = new KeyValueStoreMaterializer<>(materialized)
-                .materialize();
+            .materialize();
 
         final String source = newProcessorName(KStreamImpl.SOURCE_NAME);
         final String name = newProcessorName(KTableImpl.SOURCE_NAME);
-        final KTable<K, V> kTable = createKTable(consumed,
-                                                 topic,
-                                                 storeBuilder.name(),
-                                                 materialized.isQueryable(),
-                                                 source,
-                                                 name);
+        final ProcessorSupplier<K, V> processorSupplier = new KTableSource<>(storeBuilder.name());
+        final ProcessorParameters processorParameters = new ProcessorParameters<>(processorSupplier, name);
 
-        internalTopologyBuilder.addStateStore(storeBuilder, name);
-        internalTopologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topic);
+        StatefulSourceNode.StatefulSourceNodeBuilder<K, V> statefulSourceNodeBuilder = StatefulSourceNode.statefulSourceNodeBuilder();
 
-        return kTable;
-    }
 
+        StatefulSourceNode<K, V> statefulSourceNode = statefulSourceNodeBuilder.withNodeName(name)
+                                                                               .withSourceName(source)
+                                                                               .withStoreBuilder(storeBuilder)
+                                                                               .withConsumedInternal(consumed)
+                                                                               .withProcessorParameters(processorParameters)
+                                                                               .withTopic(topic)
+                                                                               .build();
 
-    private <K, V> KTable<K, V> createKTable(final ConsumedInternal<K, V> consumed,
-                                             final String topic,
-                                             final String storeName,
-                                             final boolean isQueryable,
-                                             final String source,
-                                             final String name) {
-        final ProcessorSupplier<K, V> processorSupplier = new KTableSource<>(storeName);
+        root.addChildNode(statefulSourceNode);
+        statefulSourceNode.setParentNode(root);
```

Comment:
I'll merge the `setParentNode` call into the `addChildNode`, but at this point, all source nodes are created in the InternalStreamsBuilder since we pass in the newly created graph node with the parent relationship already established, and in the `AbstractStream` we don't have access to the original parent node.